File: tests/typesafety/python/payment_states_misuse.py
# expect: fails
# title: receipt_id on a Pending payment, rejected by the checker
# misuse-of: payment_states
from dataclasses import dataclass


@dataclass(frozen=True)
class Pending:
    requested_at: str


@dataclass(frozen=True)
class Settled:
    requested_at: str
    receipt_id: str


type Payment = Pending | Settled


def receipt_of(payment: Payment) -> str:
    return payment.receipt_id  # error: Pending has no receipt_id

File: tests/typesafety/typescript/payment_states_misuse.ts
// expect: fails

type Pending = { readonly kind: "pending"; readonly requestedAt: string };
type Settled = { readonly kind: "settled"; readonly requestedAt: string; readonly receiptId: string };
type Payment = Pending | Settled;

export function receiptOf(payment: Payment): string {
  return payment.receiptId;
}
